src/transcript.rs

```diff
@@ -77,75 +85,89 @@ impl<R: Read, C: CurveAffine> TranscriptRead<C> for DummyHashRead<R, C> {
     fn read_scalar(&mut self) -> io::Result<C::Scalar> {
         let mut data = [0u8; 32];
         self.reader.read_exact(&mut data)?;
-        let scalar = Option::from(C::Scalar::from_bytes(&data)).ok_or_else(|| {
+        let scalar: C::Scalar = Option::from(C::Scalar::from_bytes(&data)).ok_or_else(|| {
             io::Error::new(
                 io::ErrorKind::Other,
                 "invalid field element encoding in proof",
             )
         })?;
-        self.scalar_state += &(scalar * &C::Scalar::ZETA);
-        self.scalar_state = self.scalar_state.square();
+        self.scalar_state.update(&scalar.to_bytes());
         self.read_scalar = true;
 
         Ok(scalar)
     }
 }
 
-impl<R: Read, C: CurveAffine> Transcript<C> for DummyHashRead<R, C> {
+impl<R: Read, C: CurveAffine> Transcript<C> for Blake2bRead<R, C> {
     fn common_point(&mut self, point: C) -> io::Result<()> {
         let (x, y) = Option::from(point.get_xy()).ok_or_else(|| {
             io::Error::new(
                 io::ErrorKind::Other,
                 "cannot write points at infinity to the transcript",
             )
         })?;
-        self.base_state += &(x * &C::Base::ZETA);
-        self.base_state = self.base_state.square();
-        self.base_state += &(y * &C::Base::ZETA);
-        self.base_state = self.base_state.square();
+        self.base_state.update(&x.to_bytes());
+        self.base_state.update(&y.to_bytes());
 
         Ok(())
     }
 
     fn squeeze_challenge(&mut self) -> C::Base {
```

Comment:
The way this method should work should be a bit simpler (especially after you get rid of `scalar_state`). Something like this:

```rust
let mut hasher = self.base_state.clone();
let result: [u8; 64] = hasher.finalize().as_bytes().try_into().unwrap(); 
self.base_state.update(&result[..]);
return C::Base::from_bytes_wide(&result);
```

`from_bytes_wide` allows you to avoid the loop thing while keeping the value well-distributed.